File: backend/engine/link_analyzer.py
"""
Link Analyzer
Extracts and inspects URLs from email body, checking for IP URLs, typosquatting,
credential paths, URL shorteners, punycode/homograph attacks, and suspicious TLDs.
"""

import re
from urllib.parse import urlparse
from data.threat_patterns import SUSPICIOUS_TLDS

def analyze_links(body: str) -> dict:
    urls = _extract_urls(body)
    results = [_analyze_url(u) for u in urls]

    overall_score = round(sum(r["riskScore"] for r in results) / len(results)) if results else 0
    high_risk_count = sum(1 for r in results if r["riskScore"] >= 60)

    return {
        "urls": results,
        "totalLinks": len(results),
        "highRiskLinks": high_risk_count,
        "overallScore": overall_score,
        "findings": _generate_link_findings(results),
    }

def _extract_urls(text: str) -> list:
    if not text:
        return []
    url_regex = r'https?://[^\s<>"\'\)\]]+'
    matches = re.findall(url_regex, text, re.IGNORECASE)
    href_regex = r'href=["\']([^"\']+)["\']'
    href_matches = re.findall(href_regex, text, re.IGNORECASE)

    all_urls = []
    for u in matches + href_matches:
        if u.startswith('http://') or u.startswith('https://'):
            if u not in all_urls:
                all_urls.append(u)
    return all_urls

def _analyze_url(url: str) -> dict:
    indicators = []
    risk_score = 0

    try:
        parsed = urlparse(url)
        hostname = (parsed.hostname or "").lower()
        path = (parsed.path or "").lower()
        full_url = url.lower()

        # 1. IP-based URL
        if re.match(r'^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$', hostname):
            indicators.append({
                "type": "IP_URL",
                "description": "URL uses raw IP address instead of domain name",
                "severity": "high"
            })
            risk_score += 40

        # 2. Suspicious TLD
        tld = hostname.split('.')[-1] if '.' in hostname else ''
        if tld in SUSPICIOUS_TLDS:
            indicators.append({
                "type": "SUSPICIOUS_TLD",
                "description": f"Suspicious top-level domain: .{tld}",
                "severity": "medium"
            })
            risk_score += 20

        # 3. URL shorteners
        shorteners = ['bit.ly', 'tinyurl.com', 't.co', 'goo.gl', 'ow.ly', 'is.gd', 'buff.ly', 'adf.ly', 'shorte.st', 'tiny.cc']
        if any(s in hostname for s in shorteners):
            indicators.append({
                "type": "URL_SHORTENER",
                "description": "Uses URL shortening service to mask true destination",
                "severity": "high"
            })
            risk_score += 30

        # 4. Excessive subdomains
        parts = hostname.split('.')
        if len(parts) >= 4:
            indicators.append({
                "type": "EXCESSIVE_SUBDOMAINS",
                "description": f"Suspicious subdomain nesting: {len(parts)-2} subdomains",
                "severity": "medium"
            })
            risk_score += 15

        # 5. Credential harvesting path keywords
        cred_paths = ['/login', '/signin', '/auth', '/verify', '/confirm', '/secure', '/account', '/password', '/update', '/validate']
        if any(p in path for p in cred_paths):
            indicators.append({
                "type": "CREDENTIAL_PATH",
                "description": f"URL path contains credential keyword: {path}",
                "severity": "medium"
            })
            risk_score += 20

        # 6. Typosquatting heuristics
        typos = _detect_typosquatting(hostname)
        for t in typos:
            indicators.append(t)
            risk_score += 35

        # 7. Unencrypted HTTP
        if parsed.scheme == 'http':
            indicators.append({
                "type": "NO_HTTPS",
                "description": "Uses unencrypted HTTP protocol",
                "severity": "low"
            })
            risk_score += 10

        # 8. Userinfo '@' in URL
        if '@' in full_url:
            indicators.append({
                "type": "AT_SIGN_URL",
                "description": "URL contains @ symbol — possible obfuscation trick",
                "severity": "high"
            })
            risk_score += 35

        # 9. Punycode
        if 'xn--' in hostname:
            indicators.append({
                "type": "PUNYCODE",
                "description": "Punycode/IDN domain detected — possible homograph spoofing",
                "severity": "high"
            })
            risk_score += 40

        return {
            "url": url,
            "hostname": hostname,
            "riskScore": min(100, risk_score),
            "indicators": indicators,
            "isHighRisk": risk_score >= 60,
        }
    except Exception:
        return {
            "url": url,
            "hostname": "",
            "riskScore": 20,
            "indicators": [{"type": "INVALID_URL", "description": "Malformed URL", "severity": "low"}],
            "isHighRisk": False,
        }

def _detect_typosquatting(hostname: str) -> list:
    brands = {
        'paypal.com': 'PayPal',
        'microsoft.com': 'Microsoft',
        'google.com': 'Google',
        'apple.com': 'Apple',
        'amazon.com': 'Amazon',
        'netflix.com': 'Netflix',
        'chase.com': 'Chase',
        'wellsfargo.com': 'Wells Fargo',
    }
    indicators = []
    base = hostname.split('.')[0]

    for legit_domain, brand in brands.items():
        legit_base = legit_domain.split('.')[0]
        if hostname == legit_domain or hostname.endswith('.' + legit_domain):
            continue
        if legit_base in hostname:
            indicators.append({
                "type": "BRAND_IN_HOSTNAME",
                "description": f'Domain contains brand name "{brand}" but is not legitimate {legit_domain}',
                "severity": "high"
            })
    return indicators

def _generate_link_findings(results: list) -> list:
    findings = []
    for r in results:
        if r.get("isHighRisk"):
            findings.append({
                "category": "High-Risk Link",
                "severity": "danger",
                "description": f"Suspicious URL detected: {r['url']}",
                "details": [ind["description"] for ind in r.get("indicators", [])]
            })
    return findings

File: frontend/src/engine/linkAnalyzer.js
/**
 * Link Analyzer
 * Extracts and analyzes URLs from email content for phishing indicators.
 */

import { SUSPICIOUS_TLDS } from '../data/threatPatterns';

/**
 * Analyze all links in email body
 */
export function analyzeLinks(body) {
  const urls = extractURLs(body);
  const results = [];

  for (const url of urls) {
    results.push(analyzeURL(url));
  }

  const overallScore = results.length > 0
    ? Math.round(results.reduce((sum, r) => sum + r.riskScore, 0) / results.length)
    : 0;

  const highRiskCount = results.filter(r => r.riskScore >= 60).length;

  return {
    urls: results,
    totalLinks: results.length,
    highRiskLinks: highRiskCount,
    overallScore,
    findings: generateLinkFindings(results),
  };
}

/**
 * Extract all URLs from text
 */
function extractURLs(text) {
  if (!text) return [];

  const urlRegex = /https?:\/\/[^\s<>"')\]]+/gi;
  const matches = text.match(urlRegex) || [];

  // Also check for href attributes
  const hrefRegex = /href=["']([^"']+)["']/gi;
  let hrefMatch;
  while ((hrefMatch = hrefRegex.exec(text)) !== null) {
    if (!matches.includes(hrefMatch[1])) {
      matches.push(hrefMatch[1]);
    }
  }

  return [...new Set(matches)];
}

/**
 * Analyze a single URL for risk indicators
 */
function analyzeURL(url) {
  const indicators = [];
  let riskScore = 0;

  try {
    const parsed = new URL(url);
    const hostname = parsed.hostname.toLowerCase();
    const path = parsed.pathname.toLowerCase();
    const fullUrl = url.toLowerCase();

    // 1. IP-based URL (no domain name)
    if (/^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/.test(hostname)) {
      indicators.push({ type: 'IP_URL', description: 'URL uses raw IP address instead of domain name', severity: 'high' });
      riskScore += 40;
    }

    // 2. Suspicious TLD
    const tld = hostname.split('.').pop();
    if (SUSPICIOUS_TLDS.includes(tld)) {
      indicators.push({ type: 'SUSPICIOUS_TLD', description: `Suspicious top-level domain: .${tld}`, severity: 'medium' });
      riskScore += 20;
    }

    // 3. URL shortener
    const shorteners = ['bit.ly', 'tinyurl.com', 't.co', 'goo.gl', 'ow.ly', 'is.gd', 'buff.ly', 'adf.ly', 'shorte.st', 'tiny.cc'];
    if (shorteners.some(s => hostname.includes(s))) {
      indicators.push({ type: 'URL_SHORTENER', description: 'Uses URL shortening service to hide true destination', severity: 'high' });
      riskScore += 30;
    }

    // 4. Excessive subdomains
    const subdomainCount = hostname.split('.').length - 2;
    if (subdomainCount >= 3) {
      indicators.push({ type: 'EXCESSIVE_SUBDOMAINS', description: `Suspicious: ${subdomainCount} subdomains used`, severity: 'medium' });
      riskScore += 15;
    }

    // 5. Typosquatting detection
    const typosquatIndicators = detectTyposquatting(hostname);
    if (typosquatIndicators.length > 0) {
      for (const t of typosquatIndicators) {
        indicators.push(t);
        riskScore += 35;
      }
    }

    // 6. Credential harvesting path patterns
    const credPaths = ['/login', '/signin', '/auth', '/verify', '/confirm', '/secure', '/account', '/password', '/update', '/validate'];
    if (credPaths.some(p => path.includes(p))) {
      indicators.push({ type: 'CREDENTIAL_PATH', description: `URL path contains login/credential keyword: ${path}`, severity: 'medium' });
      riskScore += 20;
    }

    // 7. Encoded/obfuscated characters
    if (url.includes('%') && (url.match(/%[0-9A-Fa-f]{2}/g) || []).length > 3) {
      indicators.push({ type: 'URL_ENCODING', description: 'Excessive URL encoding detected — possible obfuscation', severity: 'medium' });
      riskScore += 15;
    }

    // 8. Non-standard port
    if (parsed.port && !['80', '443', ''].includes(parsed.port)) {
      indicators.push({ type: 'NON_STANDARD_PORT', description: `Non-standard port: ${parsed.port}`, severity: 'medium' });
      riskScore += 15;
    }

    // 9. HTTP instead of HTTPS
    if (parsed.protocol === 'http:') {
      indicators.push({ type: 'NO_HTTPS', description: 'Uses unencrypted HTTP connection', severity: 'low' });
      riskScore += 10;
    }

    // 10. Data URI
    if (fullUrl.startsWith('data:')) {
      indicators.push({ type: 'DATA_URI', description: 'Data URI detected — could embed malicious content', severity: 'high' });
      riskScore += 40;
    }

    // 11. @ symbol in URL (user info attack)
    if (fullUrl.includes('@') && !fullUrl.includes('mailto:')) {
      indicators.push({ type: 'AT_SIGN_URL', description: 'URL contains @ symbol — possible URL obfuscation attack', severity: 'high' });
      riskScore += 35;
    }

    // 12. Homograph attack (mixed scripts)
    if (/xn--/.test(hostname)) {
      indicators.push({ type: 'PUNYCODE', description: 'Punycode/IDN domain detected — possible homograph attack', severity: 'high' });
      riskScore += 40;
    }

    return {
      url,
      hostname,
      riskScore: Math.min(100, riskScore),
      indicators,
      isHighRisk: riskScore >= 60,
    };
  } catch {
    return {
      url,
      hostname: '',
      riskScore: 20,
      indicators: [{ type: 'INVALID_URL', description: 'Malformed URL', severity: 'low' }],
      isHighRisk: false,
    };
  }
}

/**
 * Detect typosquatting (lookalike domains)
 */
function detectTyposquatting(hostname) {
  const indicators = [];
  const knownDomains = {
    'paypal.com': ['paypa1.com', 'paypai.com', 'paypaI.com', 'pαypal.com', 'paypal-secure.com', 'paypal.security-check.com', 'paypal-login.com'],
    'microsoft.com': ['microsft.com', 'micros0ft.com', 'microsoft-login.com', 'microsoft.support-center.com'],
    'google.com': ['g00gle.com', 'gogle.com', 'googIe.com', 'google-login.com'],
    'apple.com': ['app1e.com', 'appie.com', 'apple-id.com', 'apple.support-center.com'],
    'amazon.com': ['amaz0n.com', 'amazom.com', 'amazon-login.com'],
    'netflix.com': ['netf1ix.com', 'netflix-login.com'],
    'facebook.com': ['faceb00k.com', 'facebook-login.com'],
    'linkedin.com': ['linkedln.com', 'linkedin-login.com'],
  };

  for (const [legit, fakes] of Object.entries(knownDomains)) {
    // Direct match with known typosquats
    if (fakes.some(f => hostname.includes(f.split('.')[0]))) {
      indicators.push({
        type: 'TYPOSQUATTING',
        description: `Domain "${hostname}" appears to impersonate ${legit}`,
        severity: 'critical',
      });
    }

    // Check if domain contains brand name but is a different domain
    const brand = legit.split('.')[0];
    if (hostname.includes(brand) && !hostname.endsWith(legit)) {
      indicators.push({
        type: 'BRAND_IN_DOMAIN',
        description: `Domain contains "${brand}" but is not the legitimate ${legit}`,
        severity: 'high',
      });
    }
  }

  return indicators;
}

/**
 * Generate findings from link analysis
 */
function generateLinkFindings(results) {
  const findings = [];
  const highRisk = results.filter(r => r.isHighRisk);

  if (highRisk.length > 0) {
    findings.push({
      category: 'Malicious Links',
      severity: 'danger',
      description: `${highRisk.length} high-risk URL(s) detected in email body`,
      details: highRisk.map(r => `${r.url} — ${r.indicators.map(i => i.description).join('; ')}`),
    });
  }

  const shorteners = results.filter(r => r.indicators.some(i => i.type === 'URL_SHORTENER'));
  if (shorteners.length > 0) {
    findings.push({
      category: 'URL Obfuscation',
      severity: 'warning',
      description: `${shorteners.length} shortened URL(s) detected — true destinations are hidden`,
      details: shorteners.map(r => r.url),
    });
  }

  return findings;
}
